src/main/java/org/opensearch/security/filter/SecurityRestFilter.java

```diff
@@ -138,10 +139,13 @@ public void handleRequest(RestRequest request, RestChannel channel, NodeClient c
 
             NettyAttribute.popFrom(request, Netty4HttpRequestHeaderVerifier.CONTEXT_TO_RESTORE).ifPresent(storedContext -> {
                 // X_OPAQUE_ID will be overritten on restore - save to apply after restoring the saved context
-                final String xOpaqueId = threadContext.getHeader(Task.X_OPAQUE_ID);
+                final Map<String, String> tmpHeaders = threadContext.getHeaders();
                 storedContext.restore();
-                if (xOpaqueId != null) {
-                    threadContext.putHeader(Task.X_OPAQUE_ID, xOpaqueId);
+                for (Map.Entry<String, String> header : tmpHeaders.entrySet()) {
+                    threadContext.putHeader(header.getKey(), header.getValue());
```

Comment:
qq, will values be null for any header?

Reply:
does this need specific headers or all headers? Copying/populating all headers might be costly?

Reply:
Don't think this would have a high performance penalty.